### ISSUE TITLE
automatically save offline 

### ISSUE THREAD

**Issue by rs4231199** [TeamNewPipe/NewPipe#4195]
#### Describe the feature you want
It is very simple idea but will be very useful. If a video is being watched 2nd or 3rd time, then save it offline so that next time when the same video is being watched it can be played from offline copy.

**Comment by wb9688** [comment]
It does already cache streams (up to a certain amount). However the URL to the stream is different every time you'll try to load it, so that cache is not so useful after you restart NewPipe.

@Redirion: Do you happen to know if it's possible for ExoPlayer to use some custom ID (i.e. service name + stream ID + stream ID as introduced by TeamNewPipe/NewPipeExtractor#367) for the cache instead of simply the URL?

**Comment by Redirion** [comment]
nice work at your extractor pr @wb9688 
In think it could be very useful to work with other attributes than the short living URL. This might also help with some issues where the url became invalid.

However before exploring this option I would like to wait for the imminent release of ExoPlayer 2.12.

Meanwhile could you please re-review #3909?

**Comment by wb9688** [comment]
@Redirion: Yeah, let's wait for ExoPlayer 2.12 before perhaps implementing this feature. We should make an accompanying NewPipe PR for that NewPipeExtractor PR soon though, as it will fix e.g. the long standing #2415, but I hope I'll have time for that this week. I'll review #3909 later today. I'll also make more stream related PRs soon-ish for e.g. not having a hardcoded itag list for YouTube, supporting 360° videos, etc.

**Comment by SameenAhnaf** [comment]
@litetex This issue should be closed in favor of https://github.com/TeamNewPipe/NewPipe/issues/478#issuecomment-960959193

**Comment by litetex** [comment]
> @litetex This issue should be closed in favor of [#478 (comment)](https://github.com/TeamNewPipe/NewPipe/issues/478#issuecomment-960959193)

Having to counter this as the issue only is about replays.

However there is already a caching functionality inside NewPipe so I don't think this is required.
Also downloading all videos when you watch them the 2nd or 3rd (which might be done by accident) time will mess up your available disk space.

So I'm in general against that feature.
